evap/student/templates/student_vote.html

```diff
@@ -308,6 +308,88 @@ <h3 class="mb-3">{{ evaluation.full_name }} ({{ evaluation.course.semester.name
             }
+
+            const things_to_decide = [];
+            for (const form of document.getElementsByClassName("vote-form")) {
+                for (const row of form.getElementsByClassName("row")) {
+                    const selectables = row.getElementsByClassName("vote-selectable");
+                    things_to_decide.push(Array.from(selectables));
+                }
+            }
+
+            things_to_decide.push([document.getElementById("presave-btn"), document.getElementById("vote-submit-btn")]);
+
+            let current_row = 0;
+            let current_col = 0;
+
+            function clamp(min, val, max) {
+                if (val < min) return min;
+                else if (val > max) return max;
+                else return val;
+            }
+
+            function select_current(clickOnLabel=true) {
+                things_to_decide[current_row][current_col].focus();
+                if (clickOnLabel && things_to_decide[current_row][current_col].tagName === "LABEL")
+                    things_to_decide[current_row][current_col].click();
+            }
+
+            select_current();
+
+            const KEYCODES = {
+                "tab": 9,
+                "left": 37,
+                "right": 39,
+                "up": 38,
+                "down": 40,
+            };
+
+            document.getElementById("student-vote-form").addEventListener("keydown", e => {
+                if (things_to_decide[current_row][current_col].tagName == "TEXTAREA"
+                    && e.key != "Tab") {
+                    return;
+                }
+
+                console.log(e);
+                switch(e.key) {
+                    case "Tab":
+                        if (e.shiftKey) current_row--;
+                        else current_row++;
+                        current_row = clamp(0, current_row, things_to_decide.length - 1);
+
+                        // Focus the selected answer on the newly selected row
+                        let selected = things_to_decide[current_row].findIndex(el => el.classList.contains("active"));
+                        if (selected === -1) {
+                            selected = Math.floor((things_to_decide[current_row].length - 1) / 2);
+                        }
+                        current_col = selected;
+                        select_current();
+                        e.preventDefault();
+                        return;
+                    case "ArrowLeft":
+                        current_col--;
+                        break;
+                    case "ArrowRight":
+                        current_col++;
+                        break;
+                    case "ArrowUp":
+                    case "ArrowDown":
+                        e.preventDefault();
+                        return;
+                    default:
+                        const number_key = e.keyCode - 48;
```

Comment:
I'd prefer
```js
if ((index = parseInt(e.key) - 1) != NaN) {
    if (index < things_to_decide[current_row].length) {
        current_col = index - 1;
    }
}
```
or similar, to get the 48 out of there, and also not use keyCode (https://developer.mozilla.org/de/docs/Web/API/KeyboardEvent/keyCode)

Reply:
(if you do the clamping below anyway, it might make sense to remove the inner condition, so that if I press "6", "5" is selected, because that's probably what I wanted)

Reply:
Wow, and I was dead certain that `keyCode` was the sure thing to do here, in order to not fiddle around with weird strings and stuff 🤣. Well, that's that.

I can see how you could be annoyed if nothing happens when you press 6, just because you miscounted, but I am unsure where to stop here. How would I handle 0? In a (theoretical) scenario with 9 options, the 0 is closest to 9 in the number row, but closest to 1 / 2 on a numpad (and numericly). With 5 options, the zero is numericly really close to 1, but on the number row (which I feel like is the more intuitive tool for voting with the keyboard, since it aligns better with how the voting scale thingies look), I would expect that the 5 is selected.

What do you think?

Reply:
That is a good point, I only thought about the linear number line on a small keyboard.
With this in mind, I think ignoring any "invalid" number key is also fine.